packages/hash/backend-utils/src/search/opensearch.ts

```diff
@@ -162,4 +222,188 @@ export class OpenSearch extends DataSource implements SearchAdapter {
 
     return { hits };
   }
+
+  private async clearScrollAndReturnUndefined(cursor: string) {
+    await this.client.clear_scroll({ scroll_id: cursor });
+    return undefined;
+  }
+
+  /**
+   * Decide whether to increment the search cursor or to clear it.
+   * If there are no more pages left (we've seen all `total` hits) then clear
+   *
+   * @param hitCount number of hits in the result
+   * @param seenCount number of hits seen so far (not including the newly seen)
+   * @param total  number of hits for the whole search
+   * @param openSearchCursor cursor used to receive current result
+   * @param nextOpenSearchCursor cursor to use after the current result
+   * @returns a search cursor that contain more data or `undefined`
+   */
+  private async incrementOrClearOpenSearchCursor(params: {
+    pageSize: number;
+    hitCount: number;
+    seenCount: number;
+    total: number;
+    openSearchCursor: string;
+    nextOpenSearchCursor: string | undefined;
+  }) {
+    const {
+      hitCount,
+      nextOpenSearchCursor,
+      openSearchCursor,
+      seenCount,
+      total,
+      pageSize,
+    } = params;
+    const seenSoFar = seenCount + hitCount;
+
+    const endOfSearch =
+      hitCount === 0 ||
+      nextOpenSearchCursor === undefined ||
+      seenSoFar >= total;
+
+    if (endOfSearch) {
+      return await this.clearScrollAndReturnUndefined(openSearchCursor);
+    }
+
+    // A new cursor requires the page to have hits and a proper next cursor
```

Comment:
What's an improper next cursor?

Reply:
One that is either closed or not wrapped with the extension type we have around the raw OpenSearch cursors